typing/btype.ml

```diff
@@ -84,14 +84,12 @@ type changes =
   | Unchanged
   | Invalid
 
-let trail = s_table Weak.create 1
+let trail = s_table ref Unchanged
 
 let log_change ch =
-  match Weak.get !trail 0 with None -> ()
-  | Some r ->
-      let r' = ref Unchanged in
-      r := Change (ch, r');
-      Weak.set !trail 0 (Some r')
+  let r' = ref Unchanged in
+  !trail := Change (ch, r');
+  trail := r'
```

Comment:
This still looks odd to me. Before we would never mutate `trail` in this function, and now we do.

If I understand correctly, the previous type was `changes ref Weak.t ref`. I would therefore expect the transformed type to be `changes ref ref ref`, with the implementation

```ocaml
let log_change ch =
  let r = !!trail in
  let r' = ref Unchanged in
  r := Change (ch, r');
  !trail := r'
```

which is a direct translation of the previous code.

But I think that reusing the "outer" reference mutability may be correct (again, would @trefis confirm?). I am a bit confused by this interface; I suppose that if the type returned by `s_table` is a reference and not some abstract type, it means that the usage here is valid (it is okay for compiler code to mutate those references as usual), and so the patch proposed here is a non-obvious but reasonable simplification?

Reply:
Very good question!
I had to double check the implementation of the store to ensure that this is indeed OK, which I think it is.

`with_store` is going to record whatever reference trail last pointed on at the end of the execution, and restore that pointer whenever we reuse that same store. So: all good!

Reply:
I must say that looking at this PR gave me the impression that Local_store raises questions that deserve clarifications. For example we could have:
1. No difference between `ref` and `table`, only use the latter (more expressive) interface to reduce decision fatigue (this would have avoided the first confusion in this PR).
2. A private type `'a Local_store.slot` with a `get` function to avoid this second question. (And also `set`? Should we have it, documented as used only by Merlin, or is it actually enough to not have it at all and just use the internal scoping operators?)

Reply:
PRs welcome :)

(I think `set` is probably not a thing we want to expose)